lib/src/git_subprocess.rs

```diff
@@ -0,0 +1,759 @@
+
+        let output = parse_git_remote_show_output(output)?;
+
+        // find the HEAD branch line in the output
+        let branch_name = String::from_utf8(output.stdout)
+            .map_err(|e| {
+                GitSubprocessError::External(format!("git remote output is not utf-8: {e:?}"))
+            })
+            .map_err(GitError::Subprocess)?
+            .lines()
+            .map(|x| x.trim())
+            .find(|x| x.starts_with("HEAD branch:"))
+            .and_then(|x| x.split(" ").last().map(|y| y.trim().to_string()));
+
+        // git will output (unknown) if there is no default branch. we want it to be a
+        // none value
+        if let Some(x) = branch_name.as_deref() {
+            if x == "(unknown)" {
+                return Ok(None);
+            }
+        }
+        Ok(branch_name)
+    }
+
+    /// Push references to git
+    ///
+    /// All pushes are forced, using --force-with-lease to perform a test&set
+    /// operation on the remote repository
+    ///
+    /// Return tuple with
+    ///     1. refs that failed to push
+    ///     2. refs that succeeded to push
+    pub(crate) fn spawn_push(
+        &self,
+        remote_name: &str,
+        references: Vec<RefToPush>,
+    ) -> Result<(Vec<String>, Vec<String>), GitError> {
+        let args = {
+            let mut a = vec!["push".to_string(), "--porcelain".to_string()];
+
+            for reference in &references {
+                if let Some(refname) = reference.remote_reference() {
+                    a.push(format!(
+                        "--force-with-lease={refname}:{}",
+                        reference
+                            .expected_location
+                            .map(|x| x.to_string())
+                            .as_deref()
+                            .unwrap_or("")
+                    ));
+                }
+            }
+
+            a.extend_from_slice(&["--".to_string(), remote_name.to_string()]);
+
+            for reference in references {
+                a.push(reference.refspec.to_string());
+            }
+
+            a
+        };
+
+        let output = self.spawn(&args, Stdio::piped())?;
+
+        parse_git_push_output(output)
+    }
+}
+
+/// Generate a GitError::ExternalGitError if the stderr output was not
+/// recognizable
+fn external_git_error(stderr: &[u8]) -> GitError {
+    GitError::Subprocess(GitSubprocessError::External(format!(
+        "External git program failed:\n{}",
+        stderr.to_str_lossy()
+    )))
+}
+
+/// Parse no such remote errors output from git
+///
+/// To say this, git prints out a lot of things, but the first line is of the
+/// form:
+/// `fatal: '<branch_name>' does not appear to be a git repository`
+/// or
+/// `fatal: '<branch_name>': Could not resolve host: invalid-remote
+fn parse_no_such_remote(stderr: &[u8]) -> Result<(), String> {
+    if let Some(first_line) = stderr.lines().next() {
```

Comment:
nit: maybe you can use `.strip_prefix(b"...")` instead of `.starts_with_str()` and `.split_str()`.